Feature/tao 7050 dropdown component

oatymart opened the pull request:
Initial PR for code review, still preparing the work in taoInvalsi where it will be used.

To do:
~~Fix unit tests~~
~~Finalise styles~~
~~Updater & manifest~~

jsconan commented on views/js/ui/dropdown.js:
```diff
@@ -0,0 +1,240 @@
+        return component(dropdownSpecs, defaults)
+        .setTemplate(dropdownTpl)
+        // dropdown-specific init:
+        .on('init', function() {
```
this is where you can init the `open` state:

```
this.setState('open', this.config.isOpen);
```

If you want the state to be reflected as CSS class, you must put that in the `on('render'` instead.

jsconan commented on views/js/ui/dropdown.js:
```diff
@@ -0,0 +1,240 @@
+            .on('click', self.toggle)
+            .on('focus', self.open);
+        })
+        .init(config);
```
Also, pay attention to the indent. I know this is part of personal preferences, but this should at least respect the linter.

jsconan commented on views/js/ui/dropdown/tpl/dropdown.tpl:
```diff
@@ -0,0 +1,29 @@
+<div class="dropdown-container">
+    <div class="dropdown {{#if cls}} {{cls}}{{/if}} {{#if isOpen}} open{{/if}}" {{#if id}}id="{{id}}"{{/if}} data-control="{{id}}" role="navigation">
```
No need to test about inexistence of `cls`. Simply invoke the value. If it exists it will placed, else it will be ignored.
If you rely on the internal state management, the `open` class will be directly set.

jsconan commented on views/js/ui/dropdown.js:
```diff
@@ -0,0 +1,240 @@
+    var defaults = {
+        isOpen: false,
+        headerItem: {text: 'Menu'},    // li component or just html
+        innerItems: [], // array of li components or html items
```
Not sure the data should be part of the config.
Instead you should provide a way to set them at init, and also a bulk setter/getter.

jsconan commented on views/js/ui/dropdown/tpl/dropdown.tpl:
```diff
@@ -0,0 +1,29 @@
+        </div>
+        <ul class="dropdown-submenu plain" aria-label="submenu">
+            {{#each innerItems}}
+            {{! each list item can contain raw html, or structured data }}
```
A good idea would be to have a separate template for a single item. This way it will ease the insert of new items. The main template should contain only the placeholders

jsconan commented on views/scss/inc/_dropdown.scss:
```diff
@@ -0,0 +1,61 @@
+            text-align: left;
+        }
+        a:hover, a:focus, .a:hover, .a:focus {
+            background: #f3f7fa !important; // was: whiten($brand, 0.9) !important;
```
Colors should come from the main scheme. Also they should be provided through variables

jsconan commented on views/scss/inc/_dropdown.scss:
```diff
@@ -0,0 +1,61 @@
+            height: 1px;
+            width: 90%;
+            margin: 0.6rem auto;
+            background: #999;
```
same about colors

oatymart commented:
I've updated the code style and changed some of the architecture based on your feedback @jsconan .
No longer calling render() manually, but using jQuery in the API methods to modify the DOM.
The tests now pass.
I haven't implemented the 2-level template at this point.

oatymart commented:
I added the 2nd template for the list item.

addItem() is now called with an object that must be prepared, but the template is still flexible enough to allow custom HTML within the mandatory <li>.

jsconan commented on views/js/ui/dropdown.js:
```diff
@@ -0,0 +1,222 @@
+                .on('mouseleave', self.close);
+            }
+            else if (self.config.activatedBy === 'click') {
+                $component.on('click', self.toggle);
```
Caution, you are listening for click on the whole component. That means it will also be triggered when clicking on an item. That might result in events mess. You should rather target the exact control that should be clickable, I guess `this.controls.$toggler` or eventually `this.controls.$headerItem`.

oatymart replied:
Made the event listener more specific.

oatymart commented:
I changed the way the data is stored in the component, and added a visual test.

jsconan commented on views/js/ui/dropdown.js:
```diff
@@ -0,0 +1,228 @@
+            this.data = {
+                headerItem: this.config.headerItem || '',
+                innerItems: this.config.innerItems || []
+            };
```
Well, this was not what I have in mind. The data should be separated from config. And be managed as a private variable. Exposing it as a property allows to alterate it from the outside. The only way to deal with it should be through API: `setItems`, `addItem`, `setHeader`, `removeItem`, `clearItems`...
Roughly, something like that (obviously you should add some checks):
```javascript
dropdownFactory(config, data) {
    var api = {
        ...
        clearItems: function clearItems() {
            data.items = [];
            if (this.is('rendered')) {
                this.controls.$list.empty();
            }
            return this;
        },
        setItems: function setItems(items) {
            data.items = items;
            if (this.is('rendered')) {
                this.controls.$list.empty();
                _.forEach(items, function(item) {
                    this.controls.$list.append(itemTpl(item));
                });                
            }
            return this;
        },
        addItem: function addItem(item) {
            data.items.push(item);
            if (this.is('rendered')) {
                this.controls.$list.append(itemTpl(item));
            }
            return this;
        },
        removeItem: function removeItem(idx) {
            data.items.splice(idx, 1);
            this.setItems(data.items);
            return this;
        },
        ...
    };
    var dropdown = component(api, default)
        .on('render', function() {
            ....
            if (!_.empty(data)) {
                this.setItems(data.items);
                this.setHeader(data.header);
            }
            ...
        });
        .on(...);
    if (!_.empty(data)) {
        data = [];
    }
    _.defer(function() {
        dropdown.init(config);
    });
    return dropdown;
}
```

jsconan commented on views/js/ui/dropdown.js:
```diff
@@ -0,0 +1,259 @@
+            });
+        })
+        .on('item-click', function() {
+            self.close();
```
This is not working. The reason why is the `self` variable does not refer to `this` but to `window`... This is a drawback of the naming convention.

Since you are in the context of the component here, you should use `this.close();` instead.

jsconan commented on views/js/ui/dropdown.js:
```diff
@@ -0,0 +1,259 @@
+
+            // list item events
+            this.controls.$listContainer.on('click', 'li', function() {
+                var id = $(this).attr('id');
```
caution, depending on the event context, `this` could refer to a sub element. You should rather use `$(this).closest('li').attr('id')`  or `$(e.target).attr('id')`;
